proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,692 @@
+at the transport protocol layer, the resource graph layer is
+essentially irrelevant in that case.
+
+### Dynamic Parameters
+
+With the above property in mind, this document proposes the following
+data structures:
+- **Dynamic parameters**, which are a set of key/value pairs sent by the
+  client when subscribing to a resource.
+- **Dynamic parameter constraints**, which are a set of criteria that
+  can be used to determine whether a set of dynamic parameters matches
+  the constraints.  These constraints are part of the cache key for an
+  xDS resource (in addition to the resource name itself) on xDS servers,
+  xDS clients, and xDS caching proxies.  This provides a mechanism to
+  represent multiple variants of a given resource in a cacheable way.
+
+Both of these data structures are used in the xDS transport protocol,
+but they are not part of the resource name and therefore do not appear as
+part of the resource graph.
+
+When a client subscribes to a resource, it specifies a set of dynamic
+parameters.  In response, the server will send a resource whose dynamic
+parameter constraints match the dynamic parameters in the subscription
+request.  The client will use the dynamic parameter constraints on the
+returned resource to determine which of its subscriptions the resource is
+associated with.
+
+#### Constraints Representation
+
+Dynamic parameter constraints will be represented in protobuf form as follows:
+
+```proto
+message DynamicParameterConstraints {
+  // A list of constraints that may be combined with AND or OR semantics.
+  message ConstraintList {
+    // A constraint for a given key.
+    message Constraint {
+      message Exists {}
+      // The key to match against.
+      string key = 1;
+      // How to match.
+      oneof constraint_type {
+        // Matches this exact value.
+        string value = 2;
+        // Key is present (matches any value except for the key being absent).
+        Exists exists = 3;
+      }
+      // If set to true, the match is inverted -- i.e., the key must NOT
+      // match the specified value.
+      bool invert = 4;
+    }
+
+    enum MatchType {
+      // Default value.
+      MATCH_TYPE_UNSPECIFIED = 0;
+      // Logical AND of constraints.
+      MATCH_TYPE_AND = 1;
+      // Logical OR of constraints.
+      MATCH_TYPE_OR = 2;
+    }
+
+    // A list of key/value constraints.
+    repeated Constraint constraints = 1;
+
+    // How to match the constraints.
+    MatchType match_type = 2;
+  }
+
+  // A list of constraint lists. All constraint lists must match (i.e.,
+  // logical AND semantics).
+  repeated ConstraintList constraints = 1;
+}
+```
+
+#### Matching Behavior
+
+Note that both xDS servers and clients need to evaluate matching between
+a set of dynamic parameters and a set of constraints.  The server does
+this when deciding which variant of a given resource to return for a
+given subscription request.  When the client receives the resource from
```

Comment:
The client only needs to do matching behavior if it has multiple watches for the same resource with varying dynamic parameters. It was already mentioned that clients will generally only have one set of dynamic parameters for a resource (which I assume to be immutable), so it seems the common case is that clients don't need to do the constraint processing.

Reply:
> So I think that the best way to promote interoperability is to have clients do this check anyway, to make sure that servers are not just forgetting to send the constraints.

"forgetting to send the constraints" is pretty low on my list of concerns. If a server doesn't want the benefits of a proxy it seems like that's its prerogative and not much we can do about it. After all, it could continue using Node metadata or send empty constraints.

Similarly, you can't really "make" clients do the check, unless failing to do so would change behavior. And right now, it doesn't change the behavior. But we can say that clients are encouraged to do the check (and then we implement it in Envoy and gRPC).

The biggest piece I think is useful is to realize the check should be easier in end-clients because it isn't "select the correct subscription" but instead "does the returned resource match the client's constraints." That style sounds much less onerous/bug-prone to implement.

The biggest server bug risk seems to be _partial_ constraints, where the constraints don't fully encompass the selection logic. And I think you need a proxy before you'll find those. Non-matching constraints, by contrast, is the easy bug to notice; clients checking constraints only helps the case that is already easy.

It should be noted that a reason not to do the check in clients is to reduce ossification of the constraints. If there's a new constraint type, you can coordinate support in just servers and proxies. Although there are few client implementations, upgrading to newer versions may be a slower process and prevent some server-side optimization.

> because if the constraints are empty, they will still match.

That sounds okay to me. The server isn't participating and thus doesn't get the benefits, and it may not even be using constraints. I really don't see any value in requiring constraints to be present.